Core/Tests/Program.cs

```diff
@@ -21,124 +21,199 @@
+	/// if no string is given.
+	/// </summary>
+	public class TestInputReader
+	{
+		/// <summary>
+		/// Initializes a new instance of the <see cref="Loyc.Tests.TestInputReader"/> class.
+		/// Input is read from the console.
+		/// </summary>
+		public TestInputReader()
+			: this(null)
+		{ }
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="Loyc.Tests.TestInputReader"/> class
+		/// with the given input argument string.
+		/// </summary>
+		public TestInputReader(string inputArg)
+		{
+			InputArg = inputArg;
+			inputIndex = -1;
+		}
+
+		/// <summary>
+		/// Gets the input argument string.
+		/// </summary>
+		public string InputArg { get; private set; }
+
+		// Describes the index into the input argument
+		// string.
+		private int inputIndex;
+
+		/// <summary>
+		/// Reads a single character, which is either read
+		/// from the input argument string, or from the console.
+		/// A subzero value represents end-of-input.
+		/// </summary>
+		public int ReadChar()
+		{
+			if (InputArg != null) {
+				inputIndex++;
+
+				if (inputIndex >= InputArg.Length)
+					return -1;
+
+				return (int)InputArg[inputIndex];
+			} else {
+				ConsoleKeyInfo k;
+				Console.WriteLine((k = Console.ReadKey(true)).KeyChar);
+
+				if (k.Key == ConsoleKey.Escape
+					|| k.Key == ConsoleKey.Enter)
+					return -1;
+				else
+					return (int)k.KeyChar;
+			}
+		}
+	}
+
 	public class RunCoreTests
 	{
-		public static readonly VList<Pair<string, Action>> Menu = new VList<Pair<string, Action>>()
+		public static readonly VList<Pair<string, Func<bool>>> Menu = new VList<Pair<string, Func<bool>>>()
 		{
-			new Pair<string,Action>("Run unit tests of Loyc.Essentials.dll",  Loyc_Essentials),
-			new Pair<string,Action>("Run unit tests of Loyc.Collections.dll", Loyc_Collections),
-			new Pair<string,Action>("Run unit tests of Loyc.Syntax.dll",      Loyc_Syntax),
-			new Pair<string,Action>("Run unit tests of Loyc.Utilities.dll",   Loyc_Utilities),
+			new Pair<string,Func<bool>>("Run unit tests of Loyc.Essentials.dll",  Loyc_Essentials),
+			new Pair<string,Func<bool>>("Run unit tests of Loyc.Collections.dll", Loyc_Collections),
+			new Pair<string,Func<bool>>("Run unit tests of Loyc.Syntax.dll",	  Loyc_Syntax),
+			new Pair<string,Func<bool>>("Run unit tests of Loyc.Utilities.dll",   Loyc_Utilities),
 		};
 
 		public static void Main(string[] args)
 		{
 			// Workaround for MS bug: Assert(false) will not fire in debugger
 			Debug.Listeners.Clear();
 			Debug.Listeners.Add( new DefaultTraceListener() );
-			RunMenu(Menu);
+			if (!RunMenu(Menu, args))
+				// Let the outside world know that something
+				// went wrong by setting the exit code to
+				// '1'. This is particularly useful for
+				// automated tests (CI).
+				Environment.ExitCode = 1;
 		}
 
-		public static void RunMenu(IList<Pair<string, Action>> menu)
+		public static bool RunMenu(IList<Pair<string, Func<bool>>> menu, params string[] args)
```

Comment:
I don't think it's right to pass `args` to `RunMenu`, as parsing the command line should be `Main`'s job. I guess you could change `RunMenu` to take a `IEnumerable<char> inputs = null`, then as the first line do `inputs = inputs ?? ConsoleChars()`.